### PR TITLE
soc: silabs: Initialize oscillators and clock tree from DT, clean up init sequence

### PR DESCRIPTION
**Pull request by asmellby**
This PR cleans up the Series 2 init sequence by
* Unifying all Series 2 devices on using the device init HAL (adding EFR32MG21, all the others already used it)
  * DCDC is added to DeviceTree to be able to only initialize it if present
* Migrating oscillator init from device init HAL to clock manager HAL based on DT
* Migrating clock tree init from ad hoc board and driver config to clock manager HAL based on DT
* Separating soc.c for Series 2 from Series 0 and 1, since the init sequence is fully disjoint

### CODE REVIEW

File: boards/silabs/dev_kits/sltb010a/board.c
Comment (jerome-pouiller) [inline boards/silabs/dev_kits/sltb010a/board.c:1]: So good!

File: soc/silabs/silabs_s2/soc.c
Comment (jerome-pouiller) [inline soc/silabs/silabs_s2/soc.c:23]: Probably it does not hurt to include `sl_hfxo_manager.h` and `sl_power_manager.h` even if `!CONFIG_PM`.

File: soc/silabs/silabs_s2/soc.c
Comment (jerome-pouiller) [inline soc/silabs/silabs_s2/soc.c:40]: If you include `sl_hfxo_manager.h` and `sl_power_manager.h` unconditionally, you can change the `#ifdef CONFIG_PM` in `if (IS_ENABLED(CONFIG_PM))` and make the code easier.

File: soc/silabs/silabs_s2/soc.c
Comment (jerome-pouiller) [inline soc/silabs/silabs_s2/soc.c:34]: I suggest:
```
    if (IS_ENABLED(CONFIG_HW_HAS_SILABS_DCDC)) {
        sl_device_init_dcdc();
    }
```

File: dts/bindings/regulator/silabs,series2-dcdc.yaml
Comment (gmarull) [inline dts/bindings/regulator/silabs,series2-dcdc.yaml:40]: no descriptions in standard properties please

File: dts/bindings/regulator/silabs,series2-dcdc.yaml
Comment (gmarull) [inline dts/bindings/regulator/silabs,series2-dcdc.yaml:50]: why not milliamps? all values have a x1000 scale factor

File: soc/silabs/silabs_s2/Kconfig
Comment (gmarull) [inline soc/silabs/silabs_s2/Kconfig:14]: please avoid these redundant Kconfig options and just use DT natively in C code

File: soc/silabs/silabs_s2/config/sl_device_init_dcdc_config.h
Comment (gmarull) [inline soc/silabs/silabs_s2/config/sl_device_init_dcdc_config.h:11]: no soc.h please

File: soc/silabs/silabs_s2/config/sl_device_init_dcdc_config.h
Comment (gmarull) [inline soc/silabs/silabs_s2/config/sl_device_init_dcdc_config.h:37]: aren't enums correlated to voltages so that something like linear range can be used? This looks seriously ugly.

File: soc/silabs/silabs_s2/config/sl_device_init_dcdc_config.h
Comment (gmarull) [inline soc/silabs/silabs_s2/config/sl_device_init_dcdc_config.h:47]: same

File: boards/silabs/radio_boards/slwrb4180a/slwrb4180a_defconfig
Comment (gmarull) [inline boards/silabs/radio_boards/slwrb4180a/slwrb4180a_defconfig:8]: this setting should be grabbed from DT, see other examples in tree

File: dts/arm/silabs/efr32mg21.dtsi
Comment (gmarull) [inline dts/arm/silabs/efr32mg21.dtsi:30]: is this device addressable? if so, it should have a `@reg`.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: dts/bindings/regulator/silabs,series2-dcdc.yaml
Comment (asmellby) [reply dts/bindings/regulator/silabs,series2-dcdc.yaml:50]: I used microamp for symmetry with the standard regulator properties `regulator-*-microamp`. I agree that the scale factor of 1000 is silly, and if milliamp is OK, I will change to it.

File: soc/silabs/silabs_s2/Kconfig
Comment (asmellby) [reply soc/silabs/silabs_s2/Kconfig:14]: I had been pointed to https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nordic/common/Kconfig.peripherals in the past, but if this is a pattern that we don't want to propagate, and just use DT directly instead, that is completely fine by me :)

File: dts/bindings/clock/silabs,series2-lfrco.yaml
Comment (jerome-pouiller) [inline dts/bindings/clock/silabs,series2-lfrco.yaml:4]: Help messages are the right place to expand acronyms (in many case it provide a sufficient documentation).

File: dts/bindings/clock/silabs,series2-lfxo.yaml
Comment (jerome-pouiller) [inline dts/bindings/clock/silabs,series2-lfxo.yaml:25]: HFXO? (BTW, you can probably get rid of this description. "LFXO mode" is not very helpful when the attribute is named "mode" and the device "LFXO").

**Review by jerome-pouiller — DISMISSED** [review]
I still have some concerns with use of `config/sl_*_config.h`. I believe we should work on a long term fix for these files. However, we can probably live with that for now.

That's said, I am happy to see much ugly code disappeared with this PR :-).

**Review by jerome-pouiller — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

**Comment by jhedberg** [comment]
@gmarull could you please take another look? Thanks.

File: soc/silabs/silabs_s2/config/sl_device_init_dcdc_config.h
Comment (jhedberg) [reply soc/silabs/silabs_s2/config/sl_device_init_dcdc_config.h:9]: > shouldn;t this code be part of modules/hal_silabs? All glue code should reside in modules/

I'll let @asmellby comment on this specific SoC change, but in general we have a separate task of moving glue code out from the hal_silabs tree into `modules/hal_silabs` in the main tree (note: this directory doesn't even exist yet in the main tree). This will also include moving all Kconfig and CMake files from the `hal_silabs` tree to `modules/hal_silabs` in the main tree (i.e. we'll be adding `kconfig-ext: True` and `cmake-ext: True` to our `zephyr/module.yml`). Anyway that kind of refactoring is unrelated and out of scope for this PR.

Comment (asmellby) [reply soc/silabs/silabs_s2/config/sl_device_init_dcdc_config.h:9]: Agree that modules/ may be the better place, and that should be handled by the later refactor. I added these files in soc/silabs/ since there already was an existing config file like this there (sl_device_init_hfxo_config.h).

File: dts/arm/silabs/efr32bg2x.dtsi
Comment (asmellby) [reply dts/arm/silabs/efr32bg2x.dtsi:30]: Build will fail if invalid divisors are used. Agree that the comment may be redundant, and that separate DT bindings for every clock tree node declaring the valid enum may be the "technically correct" approach. However, it's very heavy handed, and would introduce another 20 or so bindings for Silabs. I can't see that any vendor has gone down this path before us in Zephyr. So I went for the comment as a light-weight documentation hint, as I found it better than just having the build fail.

**Review by jhedberg — DISMISSED** [review]

**Review by jerome-pouiller — DISMISSED** [review]

**Review by gmarull — CHANGES_REQUESTED** [review]
HAL glue code must go to `modules/hal_...`.

**Comment by jhedberg** [comment]
> HAL glue code must go to `modules/hal_...`.

@gmarull it would be more helpful and productive if you engaged in the in the actual discussion of this topic rather than just repeating yourself: https://github.com/zephyrproject-rtos/zephyr/pull/79415#discussion_r1799346670

**Comment by gmarull** [comment]
> > HAL glue code must go to `modules/hal_...`.
> 
> @gmarull it would be more helpful and productive if you engaged in the in the actual discussion thread for this topic rather than just repeating yourself: [#79415 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/79415#discussion_r1799346670)

This is new code, can perfectly be added to `modules/hal_silabs` right now. Code not related to this PR can of course be handled later.

**Comment by jhedberg** [comment]
> This is new code, can perfectly be added to `modules/hal_silabs` right now. Code not related to this PR can of course be handled later.

@gmarull doesn't picking up a CMakeLists.txt from there require that the corresponding module has `cmake-ext: True`?(something that's part of the other refactoring I've been referring to and what we'd really like to get for Zephyr 4.0 (however starting that works depends on this PR getting merged first).

**Comment by asmellby** [comment]
@gmarull Is my latest change what you had in mind? CMakeLists.txt will be updated to avoid referring to `modules/hal_silabs` from inside `soc/` in the follow-on PR that introduces `cmake-ext` for hal_silabs.

**Comment by asmellby** [comment]
Rebased on top of latest main to resolve conflict.

**Comment by asmellby** [comment]
Rebased on top of @jhedberg's change in https://github.com/zephyrproject-rtos/zephyr/pull/80148, adding DNM label until his PR is merged.

**Review by jhedberg — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

**Review by jerome-pouiller — APPROVED** [review]

**Review by rettichschnidi — COMMENTED** [review]
Still works for the board `sim3u1xx_dk` 👍

**Pull request merged**
